File: app/src/main/java/com/runtracker/app/ui/screens/gamification/GamificationViewModel.kt
package com.runtracker.app.ui.screens.gamification

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.runtracker.shared.data.model.*
import com.runtracker.shared.data.repository.GamificationRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class GamificationViewModel @Inject constructor(
    private val repository: GamificationRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(GamificationUiState())
    val uiState: StateFlow<GamificationUiState> = _uiState.asStateFlow()

    private val _newAchievements = MutableStateFlow<List<Achievement>>(emptyList())
    val newAchievements: StateFlow<List<Achievement>> = _newAchievements.asStateFlow()

    init {
        loadGamificationData()
        seedAchievements()
        checkForNewAchievements()
    }

    private fun seedAchievements() {
        viewModelScope.launch {
            repository.seedDefaultAchievements()
        }
    }

    private fun loadGamificationData() {
        viewModelScope.launch {
            // Load user gamification
            repository.getUserGamification().collect { gamification ->
                _uiState.update { it.copy(gamification = gamification) }
            }
        }
        
        viewModelScope.launch {
            // Load today's rings
            val rings = repository.getTodayRings()
            _uiState.update { it.copy(todayRings = rings) }
        }
        
        viewModelScope.launch {
            // Load recent rings for week view
            repository.getRecentDailyRings(7).collect { rings ->
                _uiState.update { it.copy(weeklyRings = rings) }
            }
        }
        
        viewModelScope.launch {
            // Load all achievements with progress
            combine(
                repository.getAllAchievements(),
                repository.getUserAchievements()
            ) { achievements, userAchievements ->
                achievements.map { achievement ->
                    val userProgress = userAchievements.find { it.achievementId == achievement.id }
                    AchievementWithProgress(
                        achievement = achievement,
                        progress = userProgress?.progress ?: 0,
                        isUnlocked = userProgress?.isUnlocked ?: false,
                        unlockedAt = userProgress?.unlockedAt
                    )
                }
            }.collect { achievementsWithProgress ->
                _uiState.update { it.copy(achievements = achievementsWithProgress) }
            }
        }
        
        viewModelScope.launch {
            // Load recent XP transactions
            repository.getRecentXpTransactions(20).collect { transactions ->
                _uiState.update { it.copy(recentXp = transactions) }
            }
        }
    }

    private fun checkForNewAchievements() {
        viewModelScope.launch {
            val unshown = repository.getUnshownAchievementNotifications()
            if (unshown.isNotEmpty()) {
                val achievements = _uiState.value.achievements
                    .filter { ap -> unshown.any { it.achievementId == ap.achievement.id } }
                    .map { it.achievement }
                _newAchievements.value = achievements
            }
        }
    }

    fun dismissAchievementNotification(achievementId: String) {
        viewModelScope.launch {
            repository.markAchievementNotificationShown(achievementId)
            _newAchievements.update { list -> list.filter { it.id != achievementId } }
        }
    }

    fun refreshData() {
        viewModelScope.launch {
            val rings = repository.getTodayRings()
            _uiState.update { it.copy(todayRings = rings) }
        }
    }
}

data class GamificationUiState(
    val gamification: UserGamification? = null,
    val todayRings: DailyRings? = null,
    val weeklyRings: List<DailyRings> = emptyList(),
    val achievements: List<AchievementWithProgress> = emptyList(),
    val recentXp: List<XpTransaction> = emptyList(),
    val isLoading: Boolean = false
) {
    val unlockedAchievements: List<AchievementWithProgress>
        get() = achievements.filter { it.isUnlocked }
    
    val lockedAchievements: List<AchievementWithProgress>
        get() = achievements.filter { !it.isUnlocked && !it.achievement.isSecret }
    
    val achievementsByCategory: Map<AchievementCategory, List<AchievementWithProgress>>
        get() = achievements.groupBy { it.achievement.category }
}

data class AchievementWithProgress(
    val achievement: Achievement,
    val progress: Int,
    val isUnlocked: Boolean,
    val unlockedAt: Long?
) {
    val progressPercent: Float
        get() = (progress.toFloat() / achievement.requirement).coerceIn(0f, 1f)
}

File: iosApp/GoSteady/Sources/ViewModels/GamificationViewModel.swift
import Foundation
import Observation
import SwiftData

// MARK: - ViewModel-specific Types

struct AchievementWithProgress: Identifiable {
    let achievement: Achievement
    var progress: Int
    var isUnlocked: Bool
    var unlockedAt: Date?

    var id: String { achievement.id }
    var progressPercent: Float {
        guard achievement.requirement > 0 else { return 0 }
        return min(max(Float(progress) / Float(achievement.requirement), 0), 1)
    }
}

// MARK: - ViewModel

@Observable
final class GamificationViewModel {
    private var gamificationRepository: GamificationRepository?

    var gamification: UserGamification?
    var todayRings: DailyRings?
    var weeklyRings: [DailyRings] = []
    var achievements: [AchievementWithProgress] = []
    var recentXp: [XpTransaction] = []
    var newAchievements: [Achievement] = []
    var isLoading = false

    var unlockedAchievements: [AchievementWithProgress] {
        achievements.filter(\.isUnlocked)
    }

    var lockedAchievements: [AchievementWithProgress] {
        achievements.filter { !$0.isUnlocked && !$0.achievement.isSecret }
    }

    var achievementsByCategory: [AchievementCategory: [AchievementWithProgress]] {
        Dictionary(grouping: achievements, by: { $0.achievement.category })
    }

    init() {}

    func configure(gamificationRepository: GamificationRepository) {
        self.gamificationRepository = gamificationRepository
        loadData()
    }

    func loadData() {
        guard let gamificationRepository else { return }
        isLoading = true
        gamification = gamificationRepository.fetchUserGamification()
        todayRings = gamificationRepository.getTodayRings()
        weeklyRings = gamificationRepository.fetchRecentDailyRings(days: 7)

        let allAchievements = DefaultAchievements.all
        let userAchievements = gamificationRepository.fetchUserAchievements()

        achievements = allAchievements.map { achievement in
            let userProgress = userAchievements.first(where: { $0.achievementId == achievement.id })
            return AchievementWithProgress(
                achievement: achievement,
                progress: userProgress?.progress ?? 0,
                isUnlocked: userProgress?.isUnlocked ?? false,
                unlockedAt: userProgress?.unlockedAt
            )
        }

        recentXp = gamificationRepository.fetchRecentXpTransactions(limit: 20)

        // Map unshown UserAchievements back to Achievement definitions
        let unshown = gamificationRepository.fetchUnshownNotifications()
        newAchievements = unshown.compactMap { ua in
            allAchievements.first(where: { $0.id == ua.achievementId })
        }
        isLoading = false
    }

    func dismissAchievementNotification(_ achievementId: String) {
        gamificationRepository?.markNotificationShown(achievementId: achievementId)
        newAchievements.removeAll { $0.id == achievementId }
    }

    func refreshData() {
        guard let gamificationRepository else { return }
        todayRings = gamificationRepository.getTodayRings()
    }
}

